v2/sender/sender.go

```diff
@@ -0,0 +1,83 @@
+package sender
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus"
+)
+
+// MessageBody is a type to represent that an input message body can be of any type
+type MessageBody any
+
+// SendHandlerFunction is a function that is used to modify the ServiceBus message before sending
+type SendHandlerFunction func(ctx context.Context, msg *azservicebus.Message)
+
+// CustomSender is a wrapper around the ServiceBus sender that allows for users to introduce middleware to modify the ServiceBus message before it's sent
+type CustomSender interface {
+	SendMessage(ctx context.Context, mb MessageBody, options *azservicebus.SendMessageOptions, middleware SendHandlerFunction) error
+}
+
+// Sender is an interface that allows users to substitute their own senders for the default ServiceBus sender when making a CustomSender
+type Sender interface {
+	SendMessage(ctx context.Context, message *azservicebus.Message, options *azservicebus.SendMessageOptions) error
+}
+
+// DefaultSender contains a Sender used to send the message to the ServiceBus queue and a Marshaller used to marshal any struct into a ServiceBus message
+type DefaultSender struct {
+	sender     Sender
+	marshaller Marshaller
+}
+
+var _ CustomSender = &DefaultSender{}
+
+// NewDefaultSender takes in a Sender and a Marshaller to create a new object that can send messages to the ServiceBus queue
+func NewDefaultSender(sender Sender, marshaller Marshaller) *DefaultSender {
```

Comment:
NewSender

Reply:
take options, not Marshaller directly

Reply:
I'm not sure I understand -
do you mean to take a SendOption (aka a handler/middleware)? If we do that, we don't have a guaranteed step in Send that converts the message body to a SB message, and so we can't feed the message/messagebody into the middleware chain.
Or, do you mean to abstract the Marshaller type into a separate "option" interface that has a method to marshal the messagebody into a sb message?

Reply:
Oh, looking at the v1 code I think I understand - if I understand correctly, we should have a SenderOptions field that contains a marshaller? And then construct the sender with that